numpy/core/src/umath/matmul.c.src

```diff
@@ -0,0 +1,393 @@
+#endif
+                    op  +=  os_p;
+                }
+                op  +=  os_m - (os_p * dp);
+            }
+        }
+        return;
+    }
+    for (iOuter = 0; iOuter < dOuter; iOuter++,
+                         args[0] += s0, args[1] += s1, args[2] += s2) {
+        char *ip1=args[0], *ip2=args[1], *op=args[2];
+#if @USEBLAS@ && defined(HAVE_CBLAS)
+        /*
+         * TODO: refactor this out to a inner_loop_selector, in
+         * PyUFunc_MatmulLoopSelector. But that call does not have access to
+         * n, m, p and strides.
+         */
+        if (too_big_for_blas) {
+            @TYPE@_matmul_inner_noblas(ip1, ip2, op, dm, dn, dp,
+                      is1_m, is1_n, is2_n, is2_p, os_m, os_p);
+        }
+        else if (special_case) {
+            /* Special case variants that have a 1 in the core dimensions */
+            if (scalar_out) {
+                /* row @ column, 1,1 output */
+                @TYPE@_dot(ip1, is1_n, ip2, is2_n, op, dn, NULL);
+            } else if (scalar_vec){
+                /*
+                 * 1,1d @ vector or vector @ 1,1d
+                 * could use cblas_Xaxy, but that requires 0ing output
+                 * and would not be faster (XXX prove it)
+                 */
+                @TYPE@_matmul_inner_noblas(ip1, ip2, op, dm, dn, dp,
+                          is1_m, is1_n, is2_n, is2_p, os_m, os_p);
+            } else if (vector_matrix) {
+                /* vector @ matrix, switch ip1, ip2, p and m */
+                @TYPE@_gemv((void *)ip2, is2_p, is2_n, (void *)ip1, is1_n,
+                            (void *)op, dp, dn);
+            } else if  (matrix_vector) {
+                /* matrix @ vector */
+                @TYPE@_gemv((void *)ip1, is1_m, is1_n, (void *)ip2, is2_n,
+                            (void *)op, dm, dn);
+            } else {
+                /* column @ row, 2d output, no blas needed or non-contiguous input */
+                @TYPE@_matmul_inner_noblas(ip1, ip2, op, dm, dn, dp,
+                          is1_m, is1_n, is2_n, is2_p, os_m, os_p);
+            }
+        } else {
+            /* matrix @ matrix */
+            if (i1blasable && i2blasable && o_c_blasable) {
+                @TYPE@_matmul_matrixmatrix((void *)ip1, is1_m, is1_n,
+                                           (void *)ip2, is2_n, is2_p,
+                                           (void *)op, os_m, os_p,
+                                           dm, dn, dp);
+            } else if (i1blasable && i2blasable && o_f_blasable) {
+                /*
+                 * Use transpose equivalence:
+                 * matmul(a, b, o) == matmul(b.T, a.T, o.T)
+                 */
+                @TYPE@_matmul_matrixmatrix((void *)ip2, is2_p, is2_n,
+                                           (void *)ip1, is1_n, is1_m,
+                                           (void *)op, os_p, os_m,
+                                           dm, dn, dp);
```

Comment:
I think you need to swap the `dm` and the `dp` here too

Reply:
adding test for m != p, fixing

Reply:
Also adding tests for non-contiguous output in gemv and fixing